File: k_sorted_lists/k_sorted_lists.py
"""
Return a new sorted merged list from K sorted lists, each with size N.
"""

import heapq

def sort_lists(lists):
    merged = []
    heap = []
    #heap = [[l[0], list_index, 0] for list_index, l in enumerate(lists) if len(l) != 0]
    for list_index in range(len(lists)):
        list_i = lists[list_index]
        if len(list_i) != 0:
            # value, list_index, element_index
            heap.append([list_i[0], list_index, 0])
    heapq.heapify(heap)
    
    while len(heap) != 0:
        value, list_index, element_index = heapq.heappop(heap)
        merged.append(value)
        if element_index + 1 < len(lists[list_index]):
            # Finding next minimum using heap/PQ is O(log n)
            # Finding next minimum through traverse & select is O(n)
            heapq.heappush(heap, [lists[list_index][element_index + 1],
                                  list_index,
                                  element_index + 1])
    return merged
        
def heapify(heap):
    pass

def heappop(heap):
    pass
    
lists = [[10, 15, 30], [12, 15, 20], [17, 20, 32]]
result = sort_lists(lists)
print(result)

lists = [[], [], [1,2,3]]
result = sort_lists(lists)
print(result)

File: k_sorted_lists/k_sorted_lists.cpp
// g++ -std=c++17 k_sorted_lists.cpp
/*
Return a new sorted merged list from K sorted lists, each with size N.
*/
#include <vector>
#include <queue>
#include <iostream>

struct Nums {
    int value;
    int list_index;
    int element_index;
    bool operator<(const Nums& i) const {
        return value > i.value;
    }
};

std::vector<int> sort_lists(std::vector<std::vector<int>> lists) {
    std::priority_queue<Nums> min_heap;
    for (int list_index = 0; list_index < lists.size(); list_index++) {
        std::vector<int> list = lists[list_index];
        if (!list.empty()) {
            min_heap.push(Nums{list[0], list_index, 0});
        }
    }

    std::vector<int> merged;
    while (!min_heap.empty()) {
        auto[value, list_index, element_index] = min_heap.top();
        min_heap.pop();
        merged.push_back(value);
        if (element_index + 1 < lists[list_index].size()) {
            min_heap.push(Nums{lists[list_index][element_index + 1], list_index, element_index + 1});
        }
    }
    return merged;
}

int main() {
    std::vector<std::vector<int>> lists;
    std::vector<int> result;

    lists = {{10, 15, 30}, {12, 15, 20}, {17, 20, 32}};
    result = sort_lists(lists);
    for (auto i : result) {
        std::cout << i << " ";
    }
    std::cout << std::endl;

    lists = {{}, {}, {1, 2, 3}};
    result = sort_lists(lists);
    for (auto i : result) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}